### PR TITLE
feat: option to filter hidden dir items

### LINKED ISSUE #6: Hidden files/folders
## Feature request: 
A checkbox for show or hide hidden files/folders as in Window file explorer.

### PR DESCRIPTION
**Pull request by mattm-malone**
By default, hidden items will now be filtered out in the navigator. This is the default behavior in Windows explorer and macOS finder. It is also the behavior in windows `dir` command and unix `ls` command.

For Windows, a file is hidden if the hidden attribute on an item is true.
For Unix based systems, a file is hidden if it begins with a dot (i.e.`.gitignore`).

New setting under Navigator settings:
image

Tested on both Windows & Mac.

Resolves #6

### CODE REVIEW

**Comment by aleksey-hoffman** [comment]
Thank you for the PR.
I think we need to add the option switch to both the settings page and the filter menu as you shown on the screenshot in the #6.

I need to review the PR and think this through.

Have you measured the execution time of `fsWin.getAttributesSync()`, how long does a single call take if you do the following?

```js
console.time('time::getAttributesSync')
fsWin.getAttributesSync()
console.timeEnd('time::getAttributesSync')
```

**Comment by aleksey-hoffman** [comment]
Off-topic regarding MacOS
Do you see the `youtube-dl` / `ffmpeg` / 7zip related errors in the console at launch?

I'm getting file permissions errors on MacOS. 
I think the binary directories lose their permissions after Node.js copies them into the `app storage` directory.
If you're having this issue as well, I think I'd need to run something like `chmod 700 ./app storage/bin/youtube-dl` for every bin dir after they are copied

**Comment by mattm-malone** [comment]
Measured `fsWin.getAttributesSync()` using that method for each item and got an average of 0.049 ms for each call, 2,77ms total 110 items, so not too bad at all :)

Having the setting on the filter menu as well would be good. Do you see the toggle on the settings and the filter menu being synchronized, just different places to set it? I'll throw that together later.

I'll look into that mac issue. I haven't seen the console errors but I wasn't really looking for them either.

**Comment by aleksey-hoffman** [comment]
Ok, so performance is not an issue then.

> Do you see the toggle on the settings and the filter menu being synchronized, just different places to set it?

Yeah, I think we should have them in both places for convenience. They should already synchronize automatically. The `v-model` attribute on the `v-switch` automatically gets the value from the computed property `showHiddenItems: 'storageData.settings.navigator.showHiddenItems',` and sets the switch position

**Comment by mattm-malone** [comment]
Updated PR with changes:
* "Show Hidden Files" toggle now also exists in Filter menu, as shown in #6.
  * Toggle will not show up in filter menu on Notes page since it doesn't apply to notes. 
* Tidied things up.

**Comment by mattm-malone** [comment]
Hey @aleksey-hoffman any updates on the review?

**Comment by aleksey-hoffman** [comment]
Good day @mattm-malone I will get to this PR in a few hours, 
I looked at it briefly, everything seems fine, except a few things. I will add the comments now.
 
I'm currently working on the archiver feature, adding all the supported formats.

**Comment by aleksey-hoffman** [comment]
@mattm-malone do you know if there's a way to pull a PR to VScode and see the changes with diffs (like you normally do in the "Source control" tab when you edit files locally) ?

**Comment by mattm-malone** [comment]
I've used this, I believe it gives a diff view of all PRs for a given repo inside the editor. Theoretically once the PR is checked out, if you want to unstage changes you could use `git reset --soft HEAD^`.

https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github

Also thanks for the quick reply, will be able to update later today.

**Comment by aleksey-hoffman** [comment]
Thanks mate, I will take a closer look at it later as well. 
I will include this feature into the next update

File: src/components/FilterField.vue [hunk 1]
```diff
@@ -79,6 +79,16 @@ Copyright © 2021 - present Aleksey Hoffman. All rights reserved.
                   ></v-switch>
                 </v-list-item>
 
+                <v-list-item
+                  class="px-5"
+                  dense
+                  v-if="this.$route.name !== 'notes'">
```
[the remark this answers is not in the document]
Comment (mattm-malone) [reply src/components/FilterField.vue]: Fixed

**Comment by aleksey-hoffman** [comment]
I wonder if it's worth adding this module just for setting file attributes.
Perhaps, we should just do system calls using powershell to get / set file attributes?

**Comment by aleksey-hoffman** [comment]
@mattm-malone  I wonder if it's worth adding `fswin` module just for setting file attributes.
Perhaps, we should just do system calls via system shell to get / set file attributes? Do you see any benefits using the `fswin` module instead of system shell commands?

**Comment by mattm-malone** [comment]
@aleksey-hoffman I think `fswin` might have better readability than running and parsing a command, but I see now that you have a bunch of code in place to run shell commands. I think fswin could be replaced by running `fsManager.getItemStats` and seeing if the "Hidden" attribute is there.

I have no preference either way. Going the shell route would keep the program a bit smaller and there would be one less thing to load.

**Comment by mattm-malone** [comment]
On second thought I tried it out, and using `exec` to get file attributes was _much_ slower,  varying from 8ms to 100ms per call.

**Review by aleksey-hoffman — APPROVED** [review]
Edited, resolved conflicts

**Comment by aleksey-hoffman** [comment]
> On second thought I tried it out, and using exec to get file attributes was much slower, varying from 8ms to 100ms per call.

Yeah, it's quite surprising how slow Powershell is. That's the reason why I had to create a reusable Powershell process that opens on app launch and stays opened, otherwise it adds another 500 ms per call to spawn the process itself. The process takes up ~30 MB of RAM, but what can you do. It would take a lot of time to implement the same functionality in C++.

**Comment by aleksey-hoffman** [comment]
@mattm-malone I renamed the `navigatorShowHiddenItems` property you added, and changed a few things to make sure the value is saved to storage when you toggle the option from the filter menu, so it behaves like the toggle in the settings.

I will add the ability to set different file attributes in one of the future updates. I just need to finish the permissions manager first.

Thanks for the contribution mate.

**Pull request merged**